nexus/types/src/deployment.rs

```diff
@@ -122,6 +123,10 @@ pub struct Blueprint {
     // See blueprint execution for more on this.
     pub external_dns_version: Generation,
 
+    /// the value the CockroachDB `cluster.preserve_downgrade_option` cluster
+    /// setting should be set to
```

Comment:
While I was waiting on builds last night I discovered that the default is actually an empty string (cluster settings, apparently, cannot be NULL). So I'm working on simplifying that a bit as part of wanting to have a `bp_cluster_setting` table.

Reply:
In my last commit I changed the interface to mean that:

- a None / NULL means to not touch the setting
- an Some("") means to set the setting to the empty string (in the specific case of `cluster.preserve_downgrade_version`, that means "unset it")

This should be consistent across all the code, now.

Reply:
As I was reading it the current implementation seemed fine to me, but I do wonder if we'd be better served by the kind of enum Dave suggested instead of `String`s. The empty string in particular is special value that I suspect will not really be obvious when skimming through things in the future?

Reply:
If Rust had a `NonEmptyString` that we could use in an Option-like enum and effectively mean the same thing I would probably do this, but the ability to have a type where `Default` and `SpecificVersion("")` are both possible unnerves me more than the behavior being non-obvious.

Reply:
What about the second option, where we have specific version enums and don't allow arbitrary strings at all?

Reply:
That would be an ever-growing enum, given that this is part of the blueprint and is stored in the database, right?

Although this means the hardcoded policy can be set as an enum parameter. But I'm still somewhat hesitant to encode this as an enum (and if I do so in Rust, do I also make it an enum in SQL? do we have to update the enum for each time we update cockroachdb?)

Reply:
> That would be an ever-growing enum, given that this is part of the blueprint and is stored in the database, right?
> ...
> do we have to update the enum for each time we update cockroachdb?

Yes and yes

> if I do so in Rust, do I also make it an enum in SQL? 

Not necessarily - we have many other places where the database schema technically allows states that our Rust code doesn't, so we have infallible conversions from Rust to SQL and fallible conversions from SQL to Rust. (Failing such a conversion should never happen, ofc. :sweat_smile:.) Storing a string in the DB and converting to an enum in Rust seems fine, I think?

To be clear, I'm lightly advocating for this because I find the empty-string-means-something-special a little confusing. If you think the enum with a variant-per-version's maintenance burden outweighs that, I can certainly live with the string.

Reply:
I think that if the hardcoded policy is also a variant of that enum (probably should be!), the maintenance burden won't be so high. I will experiment with this and see how it looks.